### PR TITLE
[kubernetes][parallel] hostname resolution timeout

### PR DESCRIPTION
**Pull request by valayDave**
In Kubernetes, jobsets ensure that all jobs within the set are scheduled together. However, while they may be scheduled concurrently, the order in which they start executing is not guaranteed. This can cause issues if a worker job starts before the control job, as the worker may fail to resolve the control job’s hostname, leading to a crash.

This update introduces a configurable `hostname_resolution_timeout`, allowing worker jobs to wait for a defined period to resolve the control job's hostname before terminating. This adds a layer of resilience, preventing crashes caused by timing issues between job start times.

One thing to note here is that if the control job finishes before the worker job starts, the worker will still wait for the `hostname_resolution_timeout`. Given that many higher-level decorators using @parallel already implement distributed barriers, this change ensures that the execution order does not jeopardize the success of the jobset.

### CODE REVIEW

File: metaflow/plugins/cards/card_decorator.py [hunk 1]
```diff
@@ -117,6 +117,7 @@ def step_init(
         # `'%s-%s'%(evt_name,step_name)` ensures that we capture this once per @card per @step.
         # Since there can be many steps checking if event is registered for `evt_name` will only make it check it once for all steps.
         # Hence, we have `_is_event_registered('%s-%s'%(evt_name,step_name))`
+        self._is_runtime_card = False
```
Comment (valayDave) [inline metaflow/plugins/cards/card_decorator.py:120]: Driveby change to ensure that if something crashes at the task-pre-step level then then the card decorator doesn't add an additional exceptions to the error trace. Since `self._is_runtime_card` used to be set in task-pre-step and if something fails before that executes for the card decorator then the card decorator's task finished threw one more exception saying that `_is_runtime_card`  is not present.

File: metaflow/plugins/parallel_decorator.py [hunk 2]
```diff
@@ -13,18 +13,19 @@ class ParallelDecorator(StepDecorator):
...
+                - main_ip (`str`)
                     The IP address of the control task.
-                - num_nodes : int
+                - num_nodes (`int`)
                     The total number of tasks created by @parallel
-                - node_index : int
+                - node_index (`int`)
                     The index of the current task in all the @parallel tasks.
-                - control_task_id : Optional[str]
+                - control_task_id (`Optional[str]`)
```
Comment (valayDave) [inline metaflow/plugins/parallel_decorator.py:28]: Driveby change to ensure docstring with stubs comes out readable

**Review by savingoyal — APPROVED** [review]
